### DISCUSSION TITLE
Token expire issue

### DISCUSSION THREAD

**Discussion by Sniperwolf881** [timdorr/tesla-api#628] [Q&A]
I just recently started using this library with TeslaSwift, and have faced some issues with the token after logging in.

When I print the token after logging in, I get this:

```
{
  "expires_in" : 28800,
  "refresh_token" : "xxxxxxxx",
  "id_token" : "xxxxxxxx",
  "token_type" : "Bearer",
  "access_token" : "xxxxxxxx",
  "created_at" : 1664609888.676774
}
```

I changed to `xxxxxxxx` so the code get's smaller. But, in the `expires_in` it says `28800`. If that is in seconds, that is equals to 8 hours. In my app I get logged out a few times a day. May that be the reason? And what can I do to stay logged in?

**Comment by IMgoRt** [comment]
I run a script that makes sure the 8 hours (yes 8 hours now) hasn't expired / due to expire and then use the refresh token to generate a renewed token (I also execute the same code in the event Tesla expires the token early - as it is known to do)

**Reply by Sniperwolf881** [reply]
> I run a script that makes sure the 8 hours (yes 8 hours now) hasn't expired / due to expire and then use the refresh token to generate a renewed token (I also execute the same code in the event Tesla expires the token early - as it is known to do)

So the first token expire after only 8 hours, but what about the refresh token - when does that expire?

**Reply by Sniperwolf881** [reply]
EDIT: I tried adding option to generate a refresh token using `api.refreshWebToken()`, but the new token also has `expires_in":28800`. So still stuck with having to login back in after 8 hours.

**Reply by IMgoRt** [reply]
> EDIT: I tried adding option to generate a refresh token using `api.refreshWebToken()`, but the new token also has `expires_in":28800`. So still stuck with having to login back in after 8 hours.

You're not "stuck having to log back in", you just use the refresh token every 8 hours - always save the new refresh token too, it doesn't seem to expire but my change at refresh

**Reply by Sniperwolf881** [reply]
But let's say I use the app, generate a new refresh token, and then close the app. After 10 hours I open up the app, I should still be signed in when doing as you say?

**Reply by IMgoRt** [reply]
> But let's say I use the app, generate a new refresh token, and then close the app. After 10 hours I open up the app, I should still be signed in when doing as you say?

The access token will have expired, the refresh will be valid
If you attempt to use the access token then you'll receive an expired response, just trigger your refresh script

**Reply by Sniperwolf881** [reply]
> > But let's say I use the app, generate a new refresh token, and then close the app. After 10 hours I open up the app, I should still be signed in when doing as you say?
> 
> The access token will have expired, the refresh will be valid If you attempt to use the access token then you'll receive an expired response, just trigger your refresh script

Not sure how to setup this in Swift, but I'll give it a try. This is what I currently have in my `AppDelegate`:
(Using UserDefaults just for testing purpose)
```
if let tokenString = UserDefaults.standard.object(forKey: "token_mfa") as? String,
    let tokenMapped: AuthToken = tokenString.decodeJSON(),
    let _ = UserDefaults.standard.object(forKey: "token_mfa") as? String {
    api.reuse(token: tokenMapped, email: nil)
}

return true
```

What coding language are you using?

**Reply by IMgoRt** [reply]
> > > But let's say I use the app, generate a new refresh token, and then close the app. After 10 hours I open up the app, I should still be signed in when doing as you say?
> > 
> > 
> > The access token will have expired, the refresh will be valid If you attempt to use the access token then you'll receive an expired response, just trigger your refresh script
> 
> Not sure how to setup this in Swift, but I'll give it a try. This is what I currently have in my `AppDelegate`: (Using UserDefaults just for testing purpose)
> 
> ```
> if let tokenString = UserDefaults.standard.object(forKey: "token_mfa") as? String,
>     let tokenMapped: AuthToken = tokenString.decodeJSON(),
>     let _ = UserDefaults.standard.object(forKey: "token_mfa") as? String {
>     api.reuse(token: tokenMapped, email: nil)
> }
> 
> return true
> ```
> 
> What coding language are you using?

I use javascript
My entry point into my code is to see if the token has expired - is (creation date + expires in > now), if so refresh

**Reply by Sniperwolf881** [reply]
Ok, I see.. Sad that the access token isn't valid for more then 8 hours. I have stored my access token in siri shortcuts, so I can tell siri to lock/unlock my car, but then I have to manually update it every 8 hours.

**Reply by Sniperwolf881** [reply]
@IMgoRt I have a question: If a previous token is used, `api.isAuthenticated` will not be `true`, or?

Because now I can use the token from yesterday, and still be able to lock/unlock my car, even if `api.isAuthenticated == false`.

**Reply by IMgoRt** [reply]
> @IMgoRt I have a question: If a previous token is used, `api.isAuthenticated` will not be `true`, or?
> 
> Because now I can use the token from yesterday, and still be able to lock/unlock my car, even if `api.isAuthenticated == false`.

Sorry, I don't code in that language, but if the token is past expiration then it will not be accepted and you'll get unauthorised
